enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/busted/Reference.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store.format.busted;
+
+import java.io.IOException;
+
+import org.neo4j.io.pagecache.PageCursor;
+
+import static java.lang.String.format;
+
+/**
+ * {@link #encode(long, Object, DataAdapter) Encoding} and {@link #decode(Object, DataAdapter) decoding} of {@code long}
+ * references, max 58-bit, into an as compact format as possible. Format is close to how utf-8 does similar encoding.
+ *
+ * Basically one or more header bits are used to note the number of bytes required to represent a
+ * particular {@code long} value followed by the value itself. Number of bytes used for any long ranges from
+ * 3 up to the full 8 bytes. The header bits sits in the most significant bit(s) of the most significant byte,
+ * so for that the bytes that make up a value is written (and of course read) in big-endian order.
+ *
+ * Negative values are also supported, in order to handle relative references.
+ *
+ * @author Mattias Persson
+ */
+public enum Reference
+{
+    // bit masks below contain one bit for 's' (sign) so actual address space is one bit less than advertised
+
+    // 3-byte, 23-bit addr space: 0sxx xxxx xxxx xxxx xxxx xxxx
+    BYTE_3( 3, (byte) 0b0, 1 ),
+
+    // 4-byte, 30-bit addr space: 10sx xxxx xxxx xxxx xxxx xxxx xxxx xxxx
+    BYTE_4( 4, (byte) 0b10, 2 ),
+
+    // 5-byte, 37-bit addr space: 110s xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx
+    BYTE_5( 5, (byte) 0b110, 3 ),
+
+    // 6-byte, 44-bit addr space: 1110 sxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx
+    BYTE_6( 6, (byte) 0b1110, 4 ),
+
+    // 7-byte, 51-bit addr space: 1111 0sxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx
+    BYTE_7( 7, (byte) 0b1111_0, 5 ),
+
+    // 8-byte, 59-bit addr space: 1111 1sxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx xxxx
+    BYTE_8( 8, (byte) 0b1111_1, 5 );
```

Comment:
For the record: I see this as a tailored format, not a generic one. If we need some other schema later then let's implement that specifically for that.